EIPS/eip-3521.md

```diff
@@ -0,0 +1,105 @@
+---
+eip: 3521
+title: Reduce access list cost
+author: Matt Garnett (@lightclient)
+discussions-to: https://ethereum-magicians.org/t/eip-3521-reduce-access-list-cost/6072
+status: Draft
+type: Standards Track
+category: Core
+created: 2021-04-15
+requires: 2028, 2718, 2929, 2930
+---
+
+## Simple Summary
+
+Reduce the cost of declaring `tx.to` storage keys in access lists.
+
+## Motivation
+
+Currently, a transaction must read at least 25 distinct storage slots in `tx.to`
+before it's more expensive to forego an access list.
+
+```
+ACCESS_LIST_ADDRESS_COST + (ACCESS_LIST_STORAGE_KEY_COST + WARM_STORAGE_READ_COST) * x = COLD_SLOAD_COST * x
+x = 24
+```
+
+EIP-2930 requires the address under which the storage keys reside be declared
+explicitly, since it must be addded to the EIP-2929 `accessed_addresses` list.
+However, `tx.to` is a special case that is added by default, so paying
+`ACCESS_LIST_ADDRESS_COST` for `tx.to` is essentially paying twice for the same
+address. Avoiding overpayment here will reduce the differential to just 5 unique
+reads before using an access list is cheaper -- making them a more attractive
+option.
+
+## Specification
+
+Treat the first occurence of `tx.to` in an access list as `calldata` for gas
+accounting purposes. Do not charge `ACCESS_LIST_ADDRESS_COST` for it. Storage
+keys underneath the address are unaffected.
+
+If `tx.to == nil`, `tx.to` is defined be the derived contract address created by
+the transaction.
+
+##  Rationale
+
+### Why charge at all?
+
+EIP-2930 is specifically written to make access lists simple to reason about and
+validate. It may be possible to modify the structure of the access list to avoid
+including `tx.to` explicitly, but this would renege on the spirit of EIP-2930.
+
+### Why charge as `calldata`?
+
+The cost of `calldata` was  thoroughly analyzed in EIP-2028 to determine
+a fair value that is not susceptible to denial-of-service attacks. We consider
+this the lower bound on how much transaction data should cost. Since there is
+no computation burden imposed for adding `tx.to` to the `accessed_addresses`
+map (it's added by default by EIP-2929), there is no reason to charge more than
```

Comment:
```suggestion
map (it's added by default by [EIP-2929](./eip-2929.md), there is no reason to charge more than
```
```suggestion
map (it's added by default by [EIP-2930](./eip-2930.md), there is no reason to charge more than
```
Do you mean 2930 here?

Reply:
No I do actually mean 2929 because that is the eip the default elements are added to the global access list.

Reply:
Shows how much I pay attention to EIPs... I thought 2930 introduced access lists, I didn't realize that 2929 introduced them and 2930 made them mutable by the contents of a transaction payload.  😬